### PR TITLE
Proofs fix

### PR DESCRIPTION
**Pull request by DrBlast**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -57,8 +57,8 @@
                 <artifactId>maven-compiler-plugin</artifactId>
                 <version>3.7.0</version>
                 <configuration>
-                    <source>1.6</source>
-                    <target>1.6</target>
+                    <source>8</source>
```
Comment (petermz) [inline pom.xml]: Please don't do this. We have people running WavesJ on 1.6

Comment (DrBlast) [reply pom.xml]: done

**Comment by artmur** [comment]
It is a complex feature branch with a lot of unrelated changes. It is difficult to do review for such changes. My proposal is to split changes by topics and create corresponding pull requests.

File: pom.xml [hunk 2]
```diff
@@ -95,7 +95,7 @@
                 <executions>
                     <execution>
                         <id>sign-artifacts</id>
-                        <phase>verify</phase>
+                        <phase>deploy</phase>
```
Comment (Karasiq) [inline pom.xml]: Should be `verify` to work

File: src/main/java/com/wavesplatform/wavesj/AssetBalance.java [hunk 3]
```diff
@@ -12,7 +12,7 @@
     @JsonCreator
     public AssetBalance(
             @JsonProperty("balance") Long balance,
-            @JsonProperty("assetId") String assetId) {
+            @JsonProperty("address") String assetId) {
```
Comment (Karasiq) [inline src/main/java/com/wavesplatform/wavesj/AssetBalance.java]: Why assetId is taken from address?

File: src/main/java/com/wavesplatform/wavesj/ByteUtils.java [hunk 4]
```diff
@@ -19,6 +19,14 @@
         return bytes;
     }
 
+    public static void putPaymentAsset(ByteBuffer buffer, String assetId) {
+        if (isWaves(assetId)) {
+            buffer.put((byte) 0);
+        } else {
+            buffer.put((byte) 1).putShort(toShort(Base58.decode(assetId).length)).put(Base58.decode(assetId));
```
Comment (Karasiq) [inline src/main/java/com/wavesplatform/wavesj/ByteUtils.java]: Better not to decode assetId two times

File: src/main/java/com/wavesplatform/wavesj/Hash.java [hunk 5]
```diff
@@ -25,15 +25,15 @@ private static Digest digest(ThreadLocal<Digest> cache) {
         return d;
     }
 
-    static byte[] hash(byte[] message, int ofs, int len, ThreadLocal<Digest> alg) {
+    public static byte[] hash(byte[] message, int ofs, int len, ThreadLocal<Digest> alg) {
```
Comment (Karasiq) [inline src/main/java/com/wavesplatform/wavesj/Hash.java]: Should it really be public?

File: src/main/java/com/wavesplatform/wavesj/Node.java [hunk 6]
```diff
@@ -326,6 +351,13 @@ public String data(PrivateKeyAccount from, Collection<DataEntry<?>> data, long f
         return send(Transactions.makeDataTx(from, data, fee));
     }
 
+    public String exchangeTransactio(PrivateKeyAccount from, Order buyOrder, Order sellOrder, long amount,
```
Comment (Karasiq) [inline src/main/java/com/wavesplatform/wavesj/Node.java]:
```suggestion
    public String exchangeTransaction(PrivateKeyAccount from, Order buyOrder, Order sellOrder, long amount,
```

File: src/main/java/com/wavesplatform/wavesj/Hash.java (hunk 5 above)
Comment (Karasiq) [reply src/main/java/com/wavesplatform/wavesj/Hash.java]:
```suggestion
    private static byte[] hash(byte[] message, int ofs, int len, ThreadLocal<Digest> alg) {
```

File: src/main/java/com/wavesplatform/wavesj/TransactionWithProofs.java [hunk 7]
```diff
@@ -17,7 +22,39 @@ protected void setProofs(List<ByteString> proofs) {
...
+        for (int i = newProofs.size(); i <= index; i++) {
+            newProofs.add(ByteString.EMPTY);
+        }
+        newProofs.set(index, proof);
+        return newProofs;
+    }
+
+    @Override
+    public byte[] getBytes() {
+        ByteBuffer buf = ByteBuffer.allocate(MAX_TX_SIZE);
```
Comment (Karasiq) [inline src/main/java/com/wavesplatform/wavesj/TransactionWithProofs.java]: Transaction size should be precalculated, without allocation of 10 kb

File: src/main/java/com/wavesplatform/wavesj/json/deser/OrderDeserializer.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+        Class o = null;
+
+        switch (version) {
+            case Order.V1:
+                o = OrderV1.class;
+                break;
+            case Order.V2:
+                o = OrderV2.class;
+                break;
```
Comment (Karasiq) [inline src/main/java/com/wavesplatform/wavesj/json/deser/OrderDeserializer.java]: Please add a `default` branch with an exception

File: src/main/java/com/wavesplatform/wavesj/matcher/OrderV1.java [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+            this.id = new ByteString(id);
+        } else {
+            this.id = null;
+        }
+        this.proofs = proofs;
+    }
+
+    @Override
+    public byte[] getBodyBytes() {
+        ByteBuffer buf = ByteBuffer.allocate(KBYTE);
```
Comment (Karasiq) [inline src/main/java/com/wavesplatform/wavesj/matcher/OrderV1.java]: Kilobyte pre-allocation is not necessary

File: src/main/java/com/wavesplatform/wavesj/matcher/OrderV2.java [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+        if (id != null) {
+            this.id = new ByteString(id);
+        } else {
+            this.id = null;
+        }
+    }
+
+    @Override
+    public byte[] getBodyBytes() {
+        ByteBuffer buf = ByteBuffer.allocate(KBYTE);
```
Comment (Karasiq) [inline src/main/java/com/wavesplatform/wavesj/matcher/OrderV2.java]: Kilobyte pre-allocation is not necessary

File: src/main/java/com/wavesplatform/wavesj/matcher/OrderV2.java [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+        putAsset(buf, assetPair.getAmountAsset());
+        putAsset(buf, assetPair.getPriceAsset());
+        buf.put((byte) orderType.ordinal()).putLong(price).putLong(amount)
+                .putLong(timestamp).putLong(expiration).putLong(matcherFee);
+        return ByteArraysUtils.getOnlyUsed(buf);
+    }
+
+    @Override
+    public byte[] getBytes() {
+        ByteBuffer buf = ByteBuffer.allocate(KBYTE);
```
Comment (Karasiq) [inline src/main/java/com/wavesplatform/wavesj/matcher/OrderV2.java]: Kilobyte pre-allocation is not necessary

File: src/main/java/com/wavesplatform/wavesj/transactions/AliasTransactionV1.java [hunk 12]
```diff
@@ -65,7 +65,7 @@ public long getTimestamp() {
     }
 
     @Override
-    public byte[] getBytes() {
+    public byte[] getBodyBytes() {
         ByteBuffer buf = ByteBuffer.allocate(KBYTE);
```
Comment (Karasiq) [inline src/main/java/com/wavesplatform/wavesj/transactions/AliasTransactionV1.java]: Kilobyte pre-allocation is not necessary

File: src/main/java/com/wavesplatform/wavesj/transactions/ExchangeTransactionV1.java [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+    }
+
+    @Override
+    public Order getOrder2() {
+        return sellOrder;
+    }
+
+    @Override
+    public byte[] getBodyBytes() {
+        ByteBuffer buf = ByteBuffer.allocate(KBYTE);
```
Comment (Karasiq) [inline src/main/java/com/wavesplatform/wavesj/transactions/ExchangeTransactionV1.java]: Kilobyte pre-allocation is not necessary

File: src/main/java/com/wavesplatform/wavesj/transactions/ExchangeTransactionV2.java [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+        return buyOrder;
+    }
+
+    public Order getOrder2() {
+        return sellOrder;
+    }
+
+    @Override
+    public byte[] getBodyBytes() {
+        ByteBuffer buf = ByteBuffer.allocate(KBYTE);
```
Comment (Karasiq) [inline src/main/java/com/wavesplatform/wavesj/transactions/ExchangeTransactionV2.java]: Kilobyte pre-allocation is not necessary

File: src/main/java/com/wavesplatform/wavesj/transactions/InvokeScriptTransaction.java [hunk 15]
```diff
@@ -137,13 +144,25 @@ public FunctionCall getCall() {
...
-        ByteBuffer buf = ByteBuffer.allocate(KBYTE);
+    public InvokeScriptTransaction withProof(int index, ByteString proof) {
+        List<ByteString> newProofs = updateProofs(index, proof);
+        return new InvokeScriptTransaction(chainId, senderPublicKey, dApp, call, payments, fee, feeAssetId, timestamp, newProofs);
+    }
+
+
+    @Override
+    public byte[] getBodyBytes() {
+        ByteBuffer buf = ByteBuffer.allocate(MAX_TX_SIZE);
```
Comment (Karasiq) [inline src/main/java/com/wavesplatform/wavesj/transactions/InvokeScriptTransaction.java]: 10 kb pre-allocation is not necessary

File: src/main/java/com/wavesplatform/wavesj/transactions/IssueTransactionV1.java [hunk 16]
```diff
@@ -93,8 +94,8 @@ public long getTimestamp() {
         return timestamp;
     }
 
-    public byte[] getBytes() {
-        ByteBuffer buf = ByteBuffer.allocate(10 * KBYTE);
+    public byte[] getBodyBytes() {
+        ByteBuffer buf = ByteBuffer.allocate(MAX_TX_SIZE);
```
Comment (Karasiq) [inline src/main/java/com/wavesplatform/wavesj/transactions/IssueTransactionV1.java]: 10 kb pre-allocation is not necessary

File: src/main/java/com/wavesplatform/wavesj/transactions/IssueTransactionV2.java [hunk 17]
```diff
@@ -113,7 +112,7 @@ public long getTimestamp() {
         return timestamp;
     }
 
-    public byte[] getBytes() {
+    public byte[] getBodyBytes() {
         ByteBuffer buf = ByteBuffer.allocate(10 * KBYTE);
```
Comment (Karasiq) [inline src/main/java/com/wavesplatform/wavesj/transactions/IssueTransactionV2.java]: 10 kb pre-allocation is not necessary

File: src/main/java/com/wavesplatform/wavesj/transactions/LeaseCancelTransactionV1.java [hunk 18]
```diff
@@ -59,7 +59,7 @@ public long getTimestamp() {
     }
 
     @Override
-    public byte[] getBytes() {
+    public byte[] getBodyBytes() {
         ByteBuffer buf = ByteBuffer.allocate(KBYTE);
```
Comment (Karasiq) [inline src/main/java/com/wavesplatform/wavesj/transactions/LeaseCancelTransactionV1.java]: 1 kb pre-allocation is not necessary

File: src/main/java/com/wavesplatform/wavesj/transactions/LeaseCancelTransactionV2.java [hunk 19]
```diff
@@ -68,7 +68,7 @@ public long getTimestamp() {
     }
 
     @Override
-    public byte[] getBytes() {
+    public byte[] getBodyBytes() {
         ByteBuffer buf = ByteBuffer.allocate(KBYTE);
```
Comment (Karasiq) [inline src/main/java/com/wavesplatform/wavesj/transactions/LeaseCancelTransactionV2.java]: 1 kb pre-allocation is not necessary

File: src/main/java/com/wavesplatform/wavesj/transactions/LeaseTransactionV2.java [hunk 20]
```diff
@@ -70,7 +69,7 @@ public long getTimestamp() {
     }
 
     @Override
-    public byte[] getBytes() {
+    public byte[] getBodyBytes() {
         ByteBuffer buf = ByteBuffer.allocate(KBYTE);
```
Comment (Karasiq) [inline src/main/java/com/wavesplatform/wavesj/transactions/LeaseTransactionV2.java]: 1 kb pre-allocation is not necessary

File: src/main/java/com/wavesplatform/wavesj/transactions/MassTransferTransaction.java [hunk 21]
```diff
@@ -79,7 +81,7 @@ public long getTimestamp() {
     }
 
     @Override
-    public byte[] getBytes() {
+    public byte[] getBodyBytes() {
         ByteBuffer buf = ByteBuffer.allocate(5 * KBYTE);
```
Comment (Karasiq) [inline src/main/java/com/wavesplatform/wavesj/transactions/MassTransferTransaction.java]: 5 kb pre-allocation is not necessary

File: src/main/java/com/wavesplatform/wavesj/transactions/ReissueTransactionV1.java [hunk 22]
```diff
@@ -74,7 +74,7 @@ public long getTimestamp() {
     }
 
     @Override
-    public byte[] getBytes() {
+    public byte[] getBodyBytes() {
         ByteBuffer buf = ByteBuffer.allocate(KBYTE);
```
Comment (Karasiq) [inline src/main/java/com/wavesplatform/wavesj/transactions/ReissueTransactionV1.java]: 1 kb pre-allocation is not necessary

File: src/main/java/com/wavesplatform/wavesj/transactions/ReissueTransactionV2.java [hunk 23]
```diff
@@ -86,7 +86,7 @@ public long getTimestamp() {
     }
 
     @Override
-    public byte[] getBytes() {
+    public byte[] getBodyBytes() {
         ByteBuffer buf = ByteBuffer.allocate(KBYTE);
```
Comment (Karasiq) [inline src/main/java/com/wavesplatform/wavesj/transactions/ReissueTransactionV2.java]: 1 kb pre-allocation is not necessary

File: src/main/java/com/wavesplatform/wavesj/transactions/SponsorTransaction.java [hunk 24]
```diff
@@ -68,11 +68,11 @@ public long getTimestamp() {
     }
 
     @Override
-    public byte[] getBytes() {
+    public byte[] getBodyBytes() {
         ByteBuffer buf = ByteBuffer.allocate(KBYTE);
```
Comment (Karasiq) [inline src/main/java/com/wavesplatform/wavesj/transactions/SponsorTransaction.java]: 1 kb pre-allocation is not necessary

File: pom.xml (hunk 2 above)
Comment (Karasiq) [reply pom.xml]:
```suggestion
                        <phase>verify</phase>
```

**Pull request merged**